[WIP] MessageKit 1.0 Staging Area

SD10 opened the pull request:
MessageKit 1.0:

- Remove 1 message to 1 section limitation #586
- Remove MessagesLayoutDelegate methods that size cell contents #573
- Restructure the MessageCollectionViewCell inheritance hierarchy #499
- Fix MessagesCollectionViewLayoutAttributes equality checking #575
- Add method to MessagesDisplayDelegate for async images #588
- Making cellTopLabel and cellBottomLabel have static sizes #519
- Change LabelAlignment to use NSTextAlignment and offset #556
- Refactor MessagesCollectionViewFlowLayout to use MessageSizeCalculators #583
- Create PhotoItem protocol for the associated value for MessageData.photo #582
- Improve support around custom cells #396
- Enable creating custom cell types #453
- Add .custom(Any) case to MessageData #479
- Move all cell subview origin calculations into MessagesCollectionViewCell #487
- Renaming `MessageData` to `MessageKind` #107 
- Only apply textColor for .text case of MessageData #625
- Remove explicit header and footer restriction #599
- Refactor reusable view API to be section based #615
- Remove CollectionViewReusable protocol #595

Documentation:
- [ ] MessagesCollectionViewFlowLayout
- [ ] MesagesCollectionViewLayoutAttributes
- [ ] MessageSizeCalculator
- [ ] TextMessageSizeCalculator
- [ ] MediaMessageSizeCalculator
- [ ] LocationMessageSizeCalculator
- [ ] Avatar
- [ ] AvatarPosition
- [ ] DetectorType
- [ ] LabelAlignment
- [ ] LocationMessageSnapshotOptions
- [ ] MessageData
- [ ] MessageKitDateFormatter
- [ ] MessageStyle
- [ ] Sender
- [x] LocationMessageCell
- [x] MediaMessageCell
- [x] MessageCollectionViewCell
- [x] MessageContentCell
- [x] TextMessageCell
- [ ] MessageReusableView
- [ ] InsetLabel
- [ ] MessageContainerView
- [ ] MessageLabel
- [ ] MessagesCollectionView

SD10 committed:
Remove open methods from extensions

eliburke commented:
@SD10 I only peripherally follow MK's development process, so excuse me if this was already suggested, but have you tried caching your height calculations instead of performing them with every request? 
I have a lightweight NSCache based cache that associates the cell's content object with its height and width: ``` func storedHeightForMessage(_ message: Message, cellWidth: CGFloat) -> CGFloat?```

As long as the supplied width is unchanged, the stored height is returned. If it changes during viewDidLoad/WillAppear or due to rotation or other resizing, I automatically recalculate and store the new width and height. 
I have NSAttributedStrings in my cells, but the calculations are infrequent and scrolling is quite zippy even with hundreds of messages.

SD10 commented:
@eliburke Thank you for bringing this up! This is exactly what we're trying to accomplish.

The problem is that **the width is not static** and that is something that is killing the progress/development of the project. 

The problem is the available width the `messageContainer` is based on the width of the `AvatarView`and the padding around the message (see below: both of those things are dynamic.)

We retrieve the size of these things through delegate methods:
```Swift
func avatarSize(for message: MessageType, at indexPath: IndexPath, in messagesCollectionView: MessagesCollectionView) -> CGSize

func messagePadding(for message: MessageType, at indexPath: IndexPath, in messagesCollectionView: MessagesCollectionView) -> CGSize
```

In JSQMVC, these are instance properties of the layout object. So they are static/maintain state. Any changes to these properties and we can invalidate the layout. 

***This was a mistake I made when designing the MessageKit API. Everything is so dynamic it makes caching nearly impossible.***

The only way to fix these things is making non-backward compatible API changes. Any thoughts?

SD10 committed:
Remove old deprecations

SD10 committed:
Set avatar size through properties on layout object

SD10 committed:
Set message padding through property on layout object

SD10 committed:
Set avatarposition as property on layout object

SD10 committed:
Set label positions as properties of layout object

SD10 committed:
Remove deprecations from example app

SD10 committed:
Remove cached layout attributes and cache text size

SD10 committed:
Cache calculated text size

SD10 committed:
Revert access control to internal until otherwise

SD10 committed:
Remove MessageCellLayoutContext

SD10 commented on Sources/Layout/MessagesCollectionViewFlowLayout.swift:
```diff
@@ -206,44 +473,26 @@ extension MessagesCollectionViewFlowLayout {
-        return currentLayoutContext
-    }
-}
+        guard let cachedSize = attributedTextSizeCache.object(forKey: attributedText.hash as NSNumber)?.cgSizeValue else {
```
This won't work yet. We may have seen the exact same text on another message with a different width

SD10 committed:
Create MessageCellLayout objects

SD10 committed:
Handle MessageData.custom caching

SD10 committed:
Make itemHeight lazy on intermediate attributes

SD10 committed:
Clean up some cache edge cases

SD10 committed:
Gardening and open a few methods

SD10 commented on Sources/Supporting/MessageKit+Availability.swift:
```diff
@@ -24,98 +24,40 @@
-    open var scrollsToBottomOnFirstLayout: Bool {
-        fatalError("Fatal Error: scrollsToBottomOnFirstLayout is no longer supported")
+    @available(*, deprecated: 0.14.0, message: "Removed in MessageKit 0.14.0. MessageKit no longer caches cell attributes as it leads to undefined behavior.")
+    func shouldCacheLayoutAttributes(for message: MessageType) -> Bool {
```
We can probably add this method back now

SD10 commented on Sources/Layout/MessagesCollectionViewFlowLayout.swift:
```diff
@@ -45,31 +44,51 @@ open class MessagesCollectionViewFlowLayout: UICollectionViewFlowLayout {
+    private let emojiMessageAttributesCache = NSCache<NSString, MessageIntermediateLayoutAttributes>()
+    private let photoMessageAttributesCache = NSCache<NSString, MessageIntermediateLayoutAttributes>()
+    private let videoMessageAttributesCache = NSCache<NSString, MessageIntermediateLayoutAttributes>()
+    private let locationMessageAttributesCache = NSCache<NSString, MessageIntermediateLayoutAttributes>()
```
Closer but still not right. We may only be able to handle text sizes 😕

zhongwuzw replied:
Why we define `cache` for each `messageData`? we use the `messageID` as the key, I missed something?

SD10 replied:
Well, this way we only need to invalidate the messages that changed. If you change the layout for a `.photo` message it shouldn't affect a `.text` message

zhongwuzw replied:
`.photo` and `.text` has the different `messageID`, I think it would not affect each other.

SD10 replied:
We can't just filter out all the objects in the cache that are `.photo` messages. If cache has `.text` messages and `.photo` messages. We have to remove everything. By separating them, we can invalidate all attributes for a given `MessageData` case at once.

SD10 commented:
> I have a lightweight NSCache based cache that associates the cell's content object with its height and width: func storedHeightForMessage(_ message: Message, cellWidth: CGFloat) -> CGFloat?

@eliburke Can you share your implementation for this? I want to know how you're retrieving the size from the cache using the content + width. I guess for me to do something similar I'd have to combine the hash of the cell content + the hash of the cell width. I'm not sure if XORing the two values is safe enough here 🤔

SD10 committed:
Add back shouldCacheAttributes

SD10 commented on Sources/Layout/MessagesCollectionViewFlowLayout.swift:
```diff
@@ -45,31 +44,51 @@ open class MessagesCollectionViewFlowLayout: UICollectionViewFlowLayout {
     open var messageLabelFont: UIFont {
```
It may be nice to move the `font` into the `TextMessageLayout` object.

Pros:
- Then attributed text would have a default font (see #559 and #563)
-  More intuitive API

Cons:
- Then users are required to set an emoji font, we lose the current 2x by default behavior

SD10 commented on Sources/Layout/MessagesCollectionViewFlowLayout.swift:
```diff
@@ -45,31 +44,51 @@ open class MessagesCollectionViewFlowLayout: UICollectionViewFlowLayout {
+
+    open var textMessageLayout = TextMessageLayout() {
+        didSet { textMessageAttributesCache.removeAllObjects() }
+    }
```
Not sure if I want these to be open to be honest, I don't want users removing the cache clearing behavior

SD10 replied:
We could potentially put the properties of `TextMessageLayout` and the cache into a class called `TextMessageSizeCalculator` -- similar to JSQ's `MessageBubbleSizeCalculator`. Then we can also store underlying methods in here such as `messageContainerSize` etc.

SD10 commented on Sources/Layout/MessagesCollectionViewFlowLayout.swift:
```diff
@@ -135,133 +237,326 @@ extension MessagesCollectionViewFlowLayout {
+    /// - Parameters:
+    ///   - message: The `MessageType` object for the cell.
+    ///   - indexPath: The `IndexPath` for the cell.
+    open func cellBottomLabelSize(for message: MessageType, at indexPath: IndexPath) -> CGSize {
```
Not sure if I want this open yet

SD10 commented:
@MessageKit/contributors @MessageKit/core-team it would be nice to get some eyes on this.

Essentially, we have moved the delegate methods from `MessagesLayoutDelegate` into properties of `MessagesCollectionViewFlowLayout`. This allows us to assume that the layout for a cell is static, if any changes happen to the layout configuration, we invalidate the cache using a didSet property observer.

**If you change the underlying content of a message, however, you will still have cache invalidation problems.** 

This could be:
- a) Modifying the `MessageData` of a `MessageType` ie) changing the text of a seen message
- b) Changing the `cellTopLabelText` and `cellBottomLabelText` for a cell 

In these cases, you would have to invalidate the layout attributes for the given `MessageType` using its `messageId`. 

This is a sort of "gotcha" but we would provide proper documentation and I think this is a reasonable, necessary, assumption.

Resolves #504, #573

eliburke commented:
@sd10 sorry I missed your message on this earlier. I'm not sure my actual implementation will shed much light as it is larded up with custom logic. But I can try and hit the high points, and maybe it will help with any part you haven't already built. Oh and ... go for breaking API changes. Short term pain, long term gain.

The cache indexes between a message ID and a content object: 
````     private var _cache = NSCache<NSNumber, MessageContent>() // NOTE: this is thread safe! ````
In my code, I nearly always go from IndexPath -> data source -> message object -> cache'd content object. Retrieving the content cache looks like: 

    func content(for message: Message) -> MessageContent {
        let cacheKey = NSNumber(value: message.messageId)
        if let messageMedia = _cache.object(forKey: cacheKey) {
            return messageMedia
        }
        var messageContent: MessageContent!
        if message.type == .message {
            // etc etc based on the message type
        } else {
             messageContent = UnknownMessageContent(message)
        }
        _cache.setObject(messageContent, forKey: cacheKey)
        return messageContent
     }

In ````heightForRowAt```` (or heightForCell) I then:

        let message = messages[indexPath.row] // get message
        let cellWidth = tableView.frame.width - AVATAR_WIDTH // determine width, which could vary per row
        // look for a stored height if the cellWidth is unchanged
        if let height = contentCache.storedHeightForMessage(message, cellWidth: cellWidth) {
            return height
        }
        // calculate and return a new stored height for the current width
        return contentCache.calculateHeightForMessage(message, cellWidth: cellWidth

The stored lookup function:

    func storedHeightForMessage(_ message: Message, cellWidth: CGFloat) -> CGFloat? {
        // if the width matches stored width, we have the correct height. return it
        let messageContent = content(for: message)
        if cellWidth == messageContent.cellWidth { return messageContent.cellHeight }
        return nil
    }

````calculateHeightForMessage```` calls determineHeight() for each different content type, and then saves the resulting value on the content cache object along with the width and any other cache-able values. For content that may require loading, it has variables and stub methods to fetching content. Fetching is triggered by the Table/Collection view's cellFor() call, with an embedded completion to reload the cell (if its currently displayed message is unchanged). Internally, when a fetch occurs that will change the rendering or size of the content, it just clears the stored width so that the next calls to heightFor will re-generate. 

I hope this is in some way helpful. I know it will be hard to translate to a more dynamic layout design, but I think if you are able to cache based on the message rather than the current width, it may get you further down the field towards the end zone.

SD10 commented:
@eliburke Thanks a lot for writing that up. You helped confirm that there is really little that can be done with the current dynamism of MessageKit's layout. I may be able to achieve what you described above but it may be more work than is worth it. It looks like that system assumes that the message content and the width remains unchanged. By maintaining the width as a state of the `MessagesCollectionViewLayout` object, I can confirm that width has not changed. Then we can make the assumption that message content is static, if not, users can invalidate the layout for those attributes.

TL;DR thanks for confirming some of my suspicions about the direction of this PR.

zhongwuzw commented on Sources/Layout/MessagesCollectionViewFlowLayout.swift:
```diff
@@ -45,31 +44,51 @@ open class MessagesCollectionViewFlowLayout: UICollectionViewFlowLayout {
 
-    /// The `MessageCellLayoutContext` for the current cell.
-    internal var currentLayoutContext: MessageCellLayoutContext!
+    open var customMessageLayout: MessageCellLayout? {
```
We only support one kind of custom cell?

SD10 replied:
I agree. I need to think about this

SD10 committed:
Handle layout logic using CellSizeCalculator

SD10 commented:
Jeez, this is going to be messy but I will probably have to use this PR as a staging area for 1.0. I'll merge everything into it and add CHANGELOG entries when I can, but the final result will have to be sifted for API breaking changes

SD10 committed:
Add MediaItem protocol

SD10 committed:
Add LocationItem protocol

SD10 committed:
Fix tests

SD10 committed:
Fix example app

SD10 committed:
Remove double import

SD10 committed:
Add CHANGELOG entry for #587

SD10 committed:
Remove unused String extension

SD10 committed:
Make CellSizeCalculator a protocol

SD10 committed:
Remove MessageIntermediateLayoutAttributes

SD10 committed:
Add size calculators

SD10 committed:
Add CHANGELOG entry for #579

SD10 committed:
Stop sizing cell labels to fit text

SD10 committed:
Add InsetLabel class

SD10 committed:
Modify default behavior for insets and height

SD10 committed:
Add CHANGELOG entry for #580

SD10 committed:
Remove deleted file

SD10 committed:
Add hook to configure async media on cell

SD10 committed:
Add CHANGELOG entry for #592

SD10 committed:
Improve equality checking on layout attributes

SD10 committed:
Merge pull request #592 from MessageKit/enhancement/async-media

Add hook to configure async media on cell

SD10 committed:
Merge pull request #593 from MessageKit/fix/equality

Improve equality checking on layout attributes

SD10 committed:
Refactor cell inheritance hierarchy

SD10 committed:
Refactor reusable view inheritance hierarchy

SD10 committed:
Fix failing unit tests

SD10 committed:
Add CHANGELOG entry for #596

SD10 committed:
Merge pull request #596 from MessageKit/refactor/inheritance

Refactor the cell inheritance hierarchy

SD10 committed:
Remove the limitation of 1 cell to 1 section (#601)

* Add numberOfSections delegate method

* Remove numberOfMessages(in) and add other two substitution methods

* Fix tests failure

SD10 commented:
This looks like  good indicator of what MessageKit 1.0 beta will look like.

We'll need some @MessageKit/contributors to give this a review and feedback.

zhongwuzw committed:
make calculator of layour open

SD10 committed:
Merge pull request #603 from zhongwuzw/layout-open

make calculator of layour open

zhongwuzw committed:
Add explicit access level && some comments

SD10 committed:
Merge pull request #605 from zhongwuzw/some-optimize

Add explicit access level && some comments

zhongwuzw committed:
Support custom CellSizeCalculator not restrict to MessageSizeCalculator

cwalo commented on Sources/Controllers/MessagesViewController+Menu.swift:
```diff
@@ -57,7 +57,7 @@ extension MessagesViewController {
         currentMenuController.setMenuVisible(false, animated: false)
 
-        guard let selectedCell = messagesCollectionView.cellForItem(at: selectedIndexPath) as? MessageCollectionViewCell else { return }
+        guard let selectedCell = messagesCollectionView.cellForItem(at: selectedIndexPath) as? MessageContentCell else { return }
```
I'm wondering if optionally casting to `MessageContentCell` creates a limitation that prevents subclasses of `MessageCollectionViewCell` from intercepting `menuControllerWillShow` unless you register the observer twice with a different selector.

SD10 replied:
Yeah, sounds like you're correct. I think it was a sloppy code migration. I'll fix this 👌

cwalo commented on Sources/Controllers/MessagesViewController.swift:
```diff
@@ -160,17 +160,14 @@ UICollectionViewDelegateFlowLayout, UICollectionViewDataSource {
-        let messageCount = collectionView.messagesDataSource?.numberOfMessages(in: collectionView) ?? 0
-        // There will only ever be 1 message per section
-        return messageCount > 0 ? 1 : 0
+        return collectionView.messagesDataSource?.numberOfItems(inSection: section, in: collectionView) ?? 0
```
Excited about this one!

cwalo commented on Sources/Layout/MessageSizeCalculator.swift:
```diff
@@ -0,0 +1,203 @@
+
+open class MessageSizeCalculator: CellSizeCalculator {
+
+    public var layout: UICollectionViewFlowLayout?
```
Is this just for convenience if the user subclasses and wants to access the layout?

SD10 replied:
Yeah pretty much. I use it to get the width of the cell from the layout. Not necessary depending on where your cell is

cwalo commented on Sources/Layout/MessageSizeCalculator.swift:
```diff
@@ -0,0 +1,203 @@
+    }
+
+    internal func avatarSize(for message: MessageType) -> CGSize {
+        let dataSource = messagesLayout.messagesDataSource
```
Oh, I see it gets used here. The internal, potentially fatal, getter works for me.

SD10 replied:
What do you mean?

cwalo replied:
Can disregard, made the first comment about the layout property and then saw it was actually _used_ here.

SD10 committed:
Merge pull request #606 from zhongwuzw/fix-layout

Support custom CellSizeCalculator not restrict to MessageSizeCalculator

zhongwuzw committed:
Make sizeCalculator lazy && fix access level

zhongwuzw committed:
update access control

zhongwuzw committed:
make NSConstraintLayoutSet internal

zhongwuzw committed:
Add CHANGELOG

zhongwuzw committed:
Merge pull request #607 from zhongwuzw/access-level

Make sizeCalculator lazy && fix access level

SD10 commented on Sources/Layout/LocationMessageSizeCalculator.swift:
```diff
@@ -0,0 +1,43 @@
+            }
+            return item.size
+        default:
+            fatalError("messageContainerSize received unhandled MessageDataType: \(message.data)")
```
This should go in the error namespace

SD10 commented on Sources/Layout/MediaMessageSizeCalculator.swift:
```diff
@@ -0,0 +1,48 @@
+        case .video(let item):
+            return sizeForMediaItem(maxWidth, item)
+        default:
+            fatalError("messageContainerSize received unhandled MessageDataType: \(message.data)")
```
Namespace

SD10 commented on Sources/Layout/MessageSizeCalculator.swift:
```diff
@@ -0,0 +1,203 @@
+
+import Foundation
+
+open class MessageSizeCalculator: CellSizeCalculator {
```
Need to document this class

SD10 commented on Sources/Views/InsetLabel.swift:
```diff
@@ -0,0 +1,38 @@
+
+import UIKit
+
+open class InsetLabel: UILabel {
```
Needs documentation

cwalo commented on Sources/Protocols/MediaItem.swift:
```diff
@@ -24,25 +24,14 @@
-    var messageContainerMaxWidth: CGFloat?
-    var messageContainerPadding: UIEdgeInsets?
-    var messageLabelInsets: UIEdgeInsets?
+    var placeholderImage: UIImage { get set }
```
I'm thinking it might be nice to provide a placeholder view here (e.g. activity indicator), but that can be saved for a feature PR.

SD10 replied:
Both you and @zhongwuzw have said that now. IMHO, I always thought that this was not a good practice. We would have a data model owning a view then. It results in initializing a view for each model and also modifying the view hierarchy in `cellForItem(at:)`.

cwalo commented on Sources/Protocols/MessagesDataSource.swift:
```diff
@@ -79,6 +88,10 @@ public extension MessagesDataSource {
     }
 
+    func numberOfItems(inSection section: Int, in messagesCollectionView: MessagesCollectionView) -> Int {
+        return 1
```
Pretty sure this crashes out of the box if the default is `1`.

SD10 replied:
Good catch. I bet I know where this crashes :trollface:

SD10 replied:
I think it's related to the following code, need to fix these.

https://github.com/MessageKit/MessageKit/blob/1abcf9757678d3d70671906cae468820fe7a479b/Sources/Controllers/MessagesViewController.swift#L240

https://github.com/MessageKit/MessageKit/blob/1abcf9757678d3d70671906cae468820fe7a479b/Sources/Controllers/MessagesViewController.swift#L256

SD10 replied:
Looks like I have to remove the `MessageType` arguments from the `header` and `footer` delegate methods.

 No way to resolve the `IndexPath` inside this message body now that there is no longer a `1` to `1` relationship between sections and messages.

SD10 replied:
This is such a rabbit hole. Requires me to refactor everything related to headers and footers

SD10 committed:
Document all protocols

cwalo reviewed (approved):
A few random thoughts, but no show-stoppers. LGTM 👍

SD10 committed:
Add documentation for cells

zhongwuzw committed:
Transform CellSizeCalculator from protocol to class

zhongwuzw committed:
Fix tests

SD10 committed:
Merge pull request #611 from zhongwuzw/sizeCalculator-refactor

Transform CellSizeCalculator from protocol to class

SD10 committed:
Remove MessageHeaderView and MessageFooterView

SD10 committed:
Remove CollectionViewReusable protocol

SD10 committed:
Refactor header/footer API to remove indexPath

SD10 committed:
Remove shouldDisplayHeader method

SD10 committed:
Update CHANGELOG.md

SD10 committed:
Remove failing tests

SD10 committed:
Merge pull request #615 from MessageKit/refactor/header-api

Refactor reusable view API to be section based

cwalo committed:
Make computed properties of MessagesCollectionViewFlowLayout public

This allows cellSizeCalculatorForItem to be overridden and still call super.

cwalo committed:
Move computed properties out of extension and add documentation

SD10 committed:
Merge pull request #616 from cwalo/refactor/move-open

Make computed properties of MessagesCollectionViewFlowLayout public

cwalo committed:
Only apply textColor for .text case of MessageData

cwalo committed:
Update CHANGELOG

SD10 committed:
Add name and PR #

cwalo committed:
Only apply textColor for .text case of MessageData (#625)

* Only apply textColor for .text case of MessageData

* Update CHANGELOG

* Add name and PR #

zhongwuzw committed:
Fixed font invalidation of attributedString in MessageLabel

cwalo committed:
Update access control

These are necessary to create custom cells that rely on the existing layout mechanisms.

cwalo committed:
Merge remote-tracking branch 'upstream/refactor/move-open' into refactor/move-open

# Conflicts:
#	Sources/Views/Cells/TextMessageCell.swift

cwalo committed:
Updated access control from review

SD10 committed:
Merge pull request #635 from cwalo/refactor/move-open

 Update access control

hcn1519 committed:
Fix ChatExample for SwiftLint

1. Set .swiftlint.yml for ChatExample
2. Fix Access Control for Explicit Top Level ACL Violation
3. Fix SwiftLint Warning: Colon Violation, Function Body Length
4. Fix Forced Unwrapping style

SD10 committed:
Merge pull request #638 from hcn1519/issue_553

Fix ChatExample for SwiftLint

zhongwuzw committed:
Update Example swiftlint config

SD10 commented:
I'm going to merge this and start a new branch
